File: Chapter 1/Section 1.2/gift1/gift1.cpp
/*
ID: ten.to.1
LANG: C++
TASK: gift1
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <map>

int main()
{
	std::ifstream gift1in("gift1.in");
	std::ofstream gift1out("gift1.out");
	
	//Variables
	int totalusers; gift1in >> totalusers;
	std::map<std::string, int> userbalances;
	std::vector<std::string> users;
	
	//Create Users and Balances
	std::string user;
	for(int i = 1; i <= totalusers; i++)
	{
		gift1in >> user;
		users.push_back(user);
		userbalances[user] = 0;
	}
	
	//Completing Transactions
	std::string usergiver;
	std::string currrecv;
	while(gift1in >> usergiver)
	{
		int totalmoney; gift1in >> totalmoney;
		int usersrecv; gift1in >> usersrecv;
		int giftmoney = 0;
		//Avoid division by 0
		if(usersrecv != 0)
		{
			giftmoney = totalmoney / usersrecv;
		}
		int leftmoney = totalmoney - (giftmoney * usersrecv);
		
		//Leftover Money
		userbalances[usergiver] +=  leftmoney - totalmoney;
		
		//Give Money
		for(int i = 1; i <= usersrecv; i++)
		{
			gift1in >> currrecv;
			userbalances[currrecv] += giftmoney;
		}
	}
	
	//Write to file
	for(int i = 0; i < totalusers; i++)
	{
		gift1out << users[i] << " " << userbalances[users[i]] << "\n";
	}
}

File: Chapter 1/Section 1.2/gift1/gift1.py
"""
ID: ten.to.1
LANG: PYTHON3
TASK: gift1
"""

#!/usr/bin/python
#Attempt to hardcore more values? More local?

gift1in   = open("gift1.in", "r")
gift1out  = open("gift1.out", "w")
translist = gift1in.read().split()

def CreateUsers(translist):
	users = {}
	
	#Create Users
	for i in range(1, int(translist[0]) + 1):
		users[translist[i]] = 0
	#Pop Elements
	for i in range(0, int(translist[0]) + 1):
		translist.pop(0)
	
	return users

def GiveMoney(users, translist):
	while (len(translist) != 0):
		#Deduce Amount
		curruser = translist[0]
		translist.pop(0)
		users[curruser] += -(int(translist[0]))
		
		#Give Amount
		amtofreci = int(translist[1])
		if(amtofreci != 0): giftmon = int(int(translist[0]) / amtofreci)
		keepmon = int(int(translist[0]) - giftmon * amtofreci)
		
		for i in range(0, 2):
			translist.pop(0)
		
		users[curruser] += keepmon
		
		for i in range(0, amtofreci): 
			users[translist[0]] += giftmon
			translist.pop(0)

def WriteMap(file, map):
	for i in map:
		write = str(i) + " " + str(map[i]) + "\n"
		file.write(write)

users = CreateUsers(translist)
GiveMoney(users, translist)
WriteMap(gift1out, users)

